From 9e51b4af57b99c9f5d77a44c2a18114a634a65f2 Mon Sep 17 00:00:00 2001
From: Nikita Yushchenko <nikita.yoush@cogentembedded.com>
Date: Thu, 24 Mar 2022 18:15:23 +0300
Subject: [PATCH 043/151] clk: renesas: add stubs for control domain support

This commit adds stubs to renesas-cpg-mssr driver to support control
domain's module-standby clocks.

Clients can use it by
	clocks = <&cpg CPG_CD_MOD 600>;

Each clock shall be defined (for S4) in r8a779f0-cpg-mssr.c in
r8a779f0_cd_mod_clks[] table. The numeric ID is composed from MSR_*
register index in r8a779f0_cd_mod_control_regs[] table, and bit number
inside that register.

This commit only adds two example clocks. More need to be added before
use.

Note that this is not a full-featured driver for control domain clocks.
It is just minimal code to make clients working.

Signed-off-by: Nikita Yushchenko <nikita.yoush@cogentembedded.com>
---
 .../dts/renesas/r8a779f0-ctrl-domain.dtsi     |  12 +-
 drivers/clk/renesas/r8a779f0-cpg-mssr.c       |  37 +++-
 drivers/clk/renesas/renesas-cpg-mssr.c        | 196 +++++++++++++++++-
 drivers/clk/renesas/renesas-cpg-mssr.h        |   9 +
 4 files changed, 249 insertions(+), 5 deletions(-)

diff --git a/arch/arm64/boot/dts/renesas/r8a779f0-ctrl-domain.dtsi b/arch/arm64/boot/dts/renesas/r8a779f0-ctrl-domain.dtsi
index 20e6861ac8ba..1a8e0b74c41b 100644
--- a/arch/arm64/boot/dts/renesas/r8a779f0-ctrl-domain.dtsi
+++ b/arch/arm64/boot/dts/renesas/r8a779f0-ctrl-domain.dtsi
@@ -67,6 +67,12 @@
 	};
 };
 
-
-
-
+/* To enable control comain in cpg driver, need to define second reg range.
+ *
+ * Syntax does not allow appending. And kernel code wants addresses in "reg"
+ * property. Thus have to duplicate original value and add second range.
+ */
+&cpg {
+	reg = <0 0xe6150000 0 0x4000>,
+	      <0 0xdf980000 0 0x10000>;
+};
diff --git a/drivers/clk/renesas/r8a779f0-cpg-mssr.c b/drivers/clk/renesas/r8a779f0-cpg-mssr.c
index 117ab2d7ee80..31851757be22 100644
--- a/drivers/clk/renesas/r8a779f0-cpg-mssr.c
+++ b/drivers/clk/renesas/r8a779f0-cpg-mssr.c
@@ -48,8 +48,12 @@ enum clk_ids {
 	CLK_PLL6_DIV2,
 	CLK_S0,
 	CLK_SDSRC,
+	CLK_RPCSRC,
 	CLK_OCO,
 
+	CD_CLK_HSB,
+	CD_CLK_LSB,
+
 	/* Module Clocks */
 	MOD_CLK_BASE
 };
@@ -77,6 +81,10 @@ static const struct cpg_core_clk r8a779f0_core_clks[] __initconst = {
 	DEF_FIXED(".sdsrc",	CLK_SDSRC,	CLK_PLL5_DIV2,	2, 1),
 	DEF_RATE(".oco",	CLK_OCO,	32768),
 
+	DEF_BASE(".rpcsrc",	CLK_RPCSRC,		CLK_TYPE_GEN4_RPCSRC, CLK_PLL5),
+	DEF_BASE(".rpc",	R8A779F0_CLK_RPC,	CLK_TYPE_GEN4_RPC, CLK_RPCSRC),
+	DEF_BASE("rpcd2",	R8A779F0_CLK_RPCD2,	CLK_TYPE_GEN4_RPCD2, R8A779F0_CLK_RPC),
+
 	/* Core Clock Outputs */
 	DEF_GEN4_Z("z0",	R8A779F0_CLK_Z0,	CLK_TYPE_GEN4_Z,	CLK_PLL2,	2, 0),
 	DEF_GEN4_Z("z1",	R8A779F0_CLK_Z1,	CLK_TYPE_GEN4_Z,	CLK_PLL2,	2, 0),
@@ -118,6 +126,9 @@ static const struct cpg_core_clk r8a779f0_core_clks[] __initconst = {
 
 	DEF_GEN4_OSC("osc",	R8A779F0_CLK_OSC,	CLK_EXTAL,	8),
 	DEF_GEN4_MDSEL("r",	R8A779F0_CLK_R, 29, CLK_EXTALR, 1, CLK_OCO, 1),
+
+	DEF_RATE("cd_hsb",	CD_CLK_HSB,	80 * 1000 * 1000),
+	DEF_RATE("cd_lsb",	CD_CLK_LSB,	40 * 1000 * 1000),
 };
 
 static const struct mssr_mod_clk r8a779f0_mod_clks[] __initconst = {
@@ -137,11 +148,12 @@ static const struct mssr_mod_clk r8a779f0_mod_clks[] __initconst = {
 	DEF_MOD("msi3",		621,	R8A779F0_CLK_MSO),
 	DEF_MOD("pcie0",	624,	R8A779F0_CLK_S0D2),
 	DEF_MOD("pcie1",	625,	R8A779F0_CLK_S0D2),
+	DEF_MOD("rpc",		629,	R8A779F0_CLK_RPCD2),
 	DEF_MOD("rtdm0",	630,	R8A779F0_CLK_S0D2),
 	DEF_MOD("rtdm1",	631,	R8A779F0_CLK_S0D2),
 	DEF_MOD("rtdm2",	700,	R8A779F0_CLK_S0D2),
 	DEF_MOD("rtdm3",	701,	R8A779F0_CLK_S0D2),
-<<<<<<< HEAD
+
 	DEF_MOD("scif0",	702,	R8A779F0_CLK_SASYNCPERD4),
 	DEF_MOD("scif1",	703,	R8A779F0_CLK_SASYNCPERD4),
 	DEF_MOD("scif3",	704,	R8A779F0_CLK_SASYNCPERD4),
@@ -182,6 +194,23 @@ static const struct mssr_mod_clk r8a779f0_mod_clks[] __initconst = {
 	DEF_MOD("thermal",      919,    R8A779F0_CLK_CL16M),
 };
 
+static const u16 r8a779f0_cd_mod_control_regs[] = {
+	0x1000,		/* 0 = MSR_RSCFD */
+	0x1010,		/* 1 = MSR_FLXA */
+	0x1030,		/* 2 = MSR_ETNB */
+	0x1040,		/* 3 = MSR_RSENT */
+	0x1050,		/* 4 = MSR_MSPI */
+	0x1060,		/* 5 = MSR_RLIN3 */
+	0x10f0,		/* 6 = MSR_RIIC */
+	0x1130,		/* 7 = MSR_TAUD */
+	0x1140,		/* 8 = MSR_TAUJ_ISO */
+	0x1180,		/* 9 = MSR_OSTM */
+};
+
+static const struct mssr_mod_clk r8a779f0_cd_mod_clks[] __initconst = {
+	DEF_CD_MOD("riic0",	600,	CD_CLK_LSB),
+	DEF_CD_MOD("ostm0",	900,	CD_CLK_HSB),
+};
 /*
  * CPG Clock Data
  */
@@ -236,6 +265,12 @@ const struct cpg_mssr_info r8a779f0_cpg_mssr_info __initconst = {
 	.num_mod_clks = ARRAY_SIZE(r8a779f0_mod_clks),
 	.num_hw_mod_clks = 24 * 32,
 
+	/* Control domain module clocks */
+	.cd_mod_control_regs = r8a779f0_cd_mod_control_regs,
+	.num_cd_mod_control_regs = ARRAY_SIZE(r8a779f0_cd_mod_control_regs),
+	.cd_mod_clks = r8a779f0_cd_mod_clks,
+	.num_cd_mod_clks = ARRAY_SIZE(r8a779f0_cd_mod_clks),
+
 	/* Callbacks */
 	.init = r8a779f0_cpg_mssr_init,
 	.cpg_clk_register = rcar_gen4_cpg_clk_register,
diff --git a/drivers/clk/renesas/renesas-cpg-mssr.c b/drivers/clk/renesas/renesas-cpg-mssr.c
index b6e7e6ca7ebd..6ee6c7c75620 100644
--- a/drivers/clk/renesas/renesas-cpg-mssr.c
+++ b/drivers/clk/renesas/renesas-cpg-mssr.c
@@ -172,6 +172,10 @@ struct cpg_mssr_priv {
 		u32 val;
 	} smstpcr_saved[ARRAY_SIZE(mstpsr_for_gen4)];
 
+	void __iomem *cd_base;
+	unsigned int num_cd_mod_clks;
+	const u16 *cd_mod_control_regs;
+
 	struct clk *clks[];
 };
 
@@ -277,6 +281,85 @@ static const struct clk_ops cpg_mstp_clock_ops = {
 	.is_enabled = cpg_mstp_clock_is_enabled,
 };
 
+static int cpg_cd_mstp_get_reg(struct clk_hw *hw,
+		void __iomem **out_reg, u32 *out_mask)
+{
+	struct mstp_clock *clock = to_mstp_clock(hw);
+	struct cpg_mssr_priv *priv = clock->priv;
+
+	if (!priv->cd_base)
+		return -ENXIO;
+	if (clock->index >= priv->num_cd_mod_clks)
+		return -EINVAL;
+
+	*out_reg = priv->cd_base + priv->cd_mod_control_regs[clock->index / 32];
+	*out_mask = BIT(clock->index % 32);
+	return 0;
+}
+
+static int cpg_cd_mstp_clock_endisable(struct clk_hw *hw, bool enable)
+{
+	struct cpg_mssr_priv *priv = to_mstp_clock(hw)->priv;
+	void __iomem *reg;
+	u32 mask, val;
+	unsigned long flags;
+	int ret;
+
+	ret = cpg_cd_mstp_get_reg(hw, &reg, &mask);
+	if (WARN_ON_ONCE(ret))
+		return ret;
+
+	spin_lock_irqsave(&priv->rmw_lock, flags);
+
+	/* Hardcode unlock for now */
+	writel(0xa5a5a501, priv->cd_base + 0x1710);
+
+	val = readl(reg);
+	if (enable)
+		val &= ~mask;	/* enable = clear bit */
+	else
+		val |= mask;
+	writel(val, reg);
+
+	(void)readl(reg);	/* flush write */
+
+	/* Hardcode lock for now */
+	writel(0xa5a5a500, priv->cd_base + 0x1710);
+
+	spin_unlock_irqrestore(&priv->rmw_lock, flags);
+
+	return 0;
+}
+
+static int cpg_cd_mstp_clock_enable(struct clk_hw *hw)
+{
+	return cpg_cd_mstp_clock_endisable(hw, true);
+}
+
+static void cpg_cd_mstp_clock_disable(struct clk_hw *hw)
+{
+	cpg_cd_mstp_clock_endisable(hw, false);
+}
+
+static int cpg_cd_mstp_clock_is_enabled(struct clk_hw *hw)
+{
+	void __iomem *reg;
+	u32 mask;
+	int ret;
+
+	ret = cpg_cd_mstp_get_reg(hw, &reg, &mask);
+	if (WARN_ON_ONCE(ret))
+		return 0;
+
+	return !(readl(reg) & mask);	/* bit set = disabled */
+}
+
+static const struct clk_ops cpg_cd_mstp_clock_ops = {
+	.enable = cpg_cd_mstp_clock_enable,
+	.disable = cpg_cd_mstp_clock_disable,
+	.is_enabled = cpg_cd_mstp_clock_is_enabled,
+};
+
 static
 struct clk *cpg_mssr_clk_src_twocell_get(struct of_phandle_args *clkspec,
 					 void *data)
@@ -317,6 +400,23 @@ struct clk *cpg_mssr_clk_src_twocell_get(struct of_phandle_args *clkspec,
 		clk = priv->clks[priv->num_core_clks + idx];
 		break;
 
+	case CPG_CD_MOD:
+		if (!priv->cd_base) {
+			dev_err(dev, "control domain is not available\n");
+			return ERR_PTR(-ENXIO);
+		}
+		type = "cd-module";
+		if ((clkidx % 100) >= 32 ||
+		    MOD_CLK_PACK(clkidx) >= priv->num_cd_mod_clks) {
+			dev_err(dev, "Invalid %s clock index %u\n", type,
+				clkidx);
+			return ERR_PTR(-EINVAL);
+		}
+		clk = priv->clks[priv->num_core_clks +
+				 priv->num_mod_clks +
+				 MOD_CLK_PACK(clkidx)];
+		break;
+
 	default:
 		dev_err(dev, "Invalid CPG clock type %u\n", clkspec->args[0]);
 		return ERR_PTR(-EINVAL);
@@ -478,6 +578,84 @@ static void __init cpg_mssr_register_mod_clk(const struct mssr_mod_clk *mod,
 	kfree(clock);
 }
 
+static void __init cpg_mssr_register_cd_mod_clk(const struct mssr_mod_clk *mod,
+						struct cpg_mssr_priv *priv)
+{
+	unsigned int clk_index;
+	struct clk *parent;
+	const char *parent_name;
+	struct clk_init_data mclk_init;
+	struct mstp_clock *mclk;
+	struct clk *clk;
+
+	if (!priv->cd_base)
+		return;
+
+	if (!mod->name) {
+		dev_err(priv->dev, "no cd-mod clock name\n");
+		return;
+	}
+
+	if (mod->id >= priv->num_cd_mod_clks) {
+		dev_err(priv->dev, "invalid id %d for cd-mod clock %s",
+				mod->id, mod->name);
+		return;
+	}
+	clk_index = priv->num_core_clks + priv->num_mod_clks + mod->id;
+	if (PTR_ERR(priv->clks[clk_index]) != -ENOENT) {
+		dev_err(priv->dev, "duplicate id %d for cd-mod clock %s\n",
+				mod->id, mod->name);
+		return;
+	}
+
+	if (mod->parent >= priv->num_core_clks + priv->num_mod_clks) {
+		dev_err(priv->dev, "invalid parent id %d for cd-mod clock %s\n",
+				mod->parent, mod->name);
+		return;
+	}
+	parent = priv->clks[mod->parent];
+	if (IS_ERR(parent)) {
+		dev_err(priv->dev, "no parent clock for cd-mode clock %s\n",
+				mod->name);
+		return;
+	}
+	parent_name = __clk_get_name(parent);
+
+	memset(&mclk_init, 0, sizeof(mclk_init));
+	mclk_init.name = mod->name;
+	mclk_init.ops = &cpg_cd_mstp_clock_ops;
+	mclk_init.flags = CLK_SET_RATE_PARENT;
+	mclk_init.parent_names = &parent_name;
+	mclk_init.num_parents = 1;
+
+	mclk = kzalloc(sizeof(*mclk), GFP_KERNEL);
+	if (!mclk) {
+		dev_err(priv->dev, "failed to allocate cd-mod clock %s\n",
+				mod->name);
+		return;
+	}
+	mclk->index = mod->id;
+	mclk->priv = priv;
+	mclk->hw.init = &mclk_init;
+
+	clk = clk_register(NULL, &mclk->hw);
+	if (IS_ERR(clk)) {
+		dev_err(priv->dev, "failed to register cd-mod clock %s\n",
+				mod->name);
+		goto err_register;
+	}
+
+	priv->clks[clk_index] = clk;
+	dev_info(priv->dev, "registered cd-mod clk %s, id %d, rate %ld\n",
+			mod->name,
+			(mod->id / 32 * 100) + (mod->id % 32),
+			clk_get_rate(clk));
+	return;
+
+err_register:
+	kfree(mclk);
+}
+
 struct cpg_mssr_clk_domain {
 	struct generic_pm_domain genpd;
 	unsigned int num_core_pm_clks;
@@ -972,7 +1150,8 @@ static int __init cpg_mssr_common_init(struct device *dev,
 			return error;
 	}
 
-	nclks = info->num_total_core_clks + info->num_hw_mod_clks;
+	nclks = info->num_total_core_clks + info->num_hw_mod_clks +
+			(32 * info->num_cd_mod_control_regs);
 	priv = kzalloc(struct_size(priv, clks, nclks), GFP_KERNEL);
 	if (!priv)
 		return -ENOMEM;
@@ -987,6 +1166,12 @@ static int __init cpg_mssr_common_init(struct device *dev,
 		goto out_err;
 	}
 
+	if (info->num_cd_mod_control_regs) {
+		priv->cd_base = of_iomap(np, 1);
+		/* since now, use non-null cd_base as a check that control
+		 * domain mod clocks exist */
+	}
+
 	cpg_mssr_priv = priv;
 	priv->num_core_clks = info->num_total_core_clks;
 	priv->num_mod_clks = info->num_hw_mod_clks;
@@ -1009,6 +1194,10 @@ static int __init cpg_mssr_common_init(struct device *dev,
 		error = -EINVAL;
 		goto out_err;
 	}
+	if (priv->cd_base) {
+		priv->num_cd_mod_clks = 32 * info->num_cd_mod_control_regs;
+		priv->cd_mod_control_regs = info->cd_mod_control_regs;
+	}
 
 	for (i = 0; i < nclks; i++)
 		priv->clks[i] = ERR_PTR(-ENOENT);
@@ -1022,6 +1211,8 @@ static int __init cpg_mssr_common_init(struct device *dev,
 out_err:
 	if (priv->base)
 		iounmap(priv->base);
+	if (priv->cd_base)
+		iounmap(priv->cd_base);
 	kfree(priv);
 
 	return error;
@@ -1074,6 +1265,9 @@ static int __init cpg_mssr_probe(struct platform_device *pdev)
 	for (i = 0; i < info->num_mod_clks; i++)
 		cpg_mssr_register_mod_clk(&info->mod_clks[i], info, priv);
 
+	for (i = 0; i < info->num_cd_mod_clks; i++)
+		cpg_mssr_register_cd_mod_clk(&info->cd_mod_clks[i], priv);
+
 	error = devm_add_action_or_reset(dev,
 					 cpg_mssr_del_clk_provider,
 					 np);
diff --git a/drivers/clk/renesas/renesas-cpg-mssr.h b/drivers/clk/renesas/renesas-cpg-mssr.h
index 0370e3b45553..15fe9c2e526e 100644
--- a/drivers/clk/renesas/renesas-cpg-mssr.h
+++ b/drivers/clk/renesas/renesas-cpg-mssr.h
@@ -75,6 +75,9 @@ struct mssr_mod_clk {
 #define DEF_MOD(_name, _mod, _parent...)	\
 	{ .name = _name, .id = MOD_CLK_ID(_mod), .parent = _parent }
 
+#define DEF_CD_MOD(_name, _id, _parent)		\
+	{ .name = _name, .id = MOD_CLK_PACK(_id), .parent = _parent }
+
 /* Convert from sparse base-10 to packed index space */
 #define MOD_CLK_PACK_10(x)	((x / 10) * 32 + (x % 10))
 
@@ -155,6 +158,12 @@ struct cpg_mssr_info {
 					const struct cpg_mssr_info *info,
 					struct clk **clks, void __iomem *base,
 					struct raw_notifier_head *notifiers);
+
+	/* Support for control domain module clocks */
+	const u16 *cd_mod_control_regs;
+	unsigned int num_cd_mod_control_regs;
+	const struct mssr_mod_clk *cd_mod_clks;
+	unsigned int num_cd_mod_clks;
 };
 
 extern const struct cpg_mssr_info r7s9210_cpg_mssr_info;
-- 
2.17.1

